src/client/datascience/data-viewing/dataViewer.ts

```diff
@@ -99,23 +114,31 @@ export class DataViewer extends WebviewPanelHost<IDataViewerMapping> implements
     protected onMessage(message: string, payload: any) {
         switch (message) {
             case DataViewerMessages.GetAllRowsRequest:
-                this.getAllRows().ignoreErrors();
+                this.getAllRows(payload as string).ignoreErrors();
                 break;
 
             case DataViewerMessages.GetRowsRequest:
                 this.getRowChunk(payload as IGetRowsRequest).ignoreErrors();
                 break;
 
+            case DataViewerMessages.GetSliceRequest:
+                this.getSlice(payload as IGetSliceRequest).ignoreErrors();
+                break;
+
             default:
                 break;
         }
 
         super.onMessage(message, payload);
     }
 
-    private getDataFrameInfo(): Promise<IDataFrameInfo> {
-        if (!this.dataFrameInfoPromise) {
-            this.dataFrameInfoPromise = this.dataProvider ? this.dataProvider.getDataFrameInfo() : Promise.resolve({});
+    private getDataFrameInfo(sliceExpression?: string): Promise<IDataFrameInfo> {
+        // If requesting a new slice, refresh our cached info promise
+        if (!this.dataFrameInfoPromise || sliceExpression !== this.currentSlice) {
+            this.dataFrameInfoPromise = this.dataProvider
+                ? this.dataProvider.getDataFrameInfo(sliceExpression)
+                : Promise.resolve({});
+            this.currentSlice = sliceExpression;
```

Comment:
Need to check the code, but what happens if the slice expression is invalid?

Reply:
Not a problem yet, as we're not shipping the UI control I have implemented as is. Depending on the UI (e.g. dropdowns vs a freeform input box) I'll solve this in part 2.

Reply:
Right now if the slice expression is invalid the Python scripts don't return anything. But as I mentioned the intent is to ensure that our slice expression is always valid, e.g. through dropdowns or validating the provided slice expression against the variable's shape